### PR TITLE
treat items in queues as implicit tasks; queues to be list separated type

### LINKED ISSUE #5260: Space-separated list of queue members ignored without error
**Describe the bug**
See [discussion on Element](https://matrix.to/#/!hwZqSYihGPuhDdIzIP:matrix.org/$1670255552138977aoIyT:matrix.org?via=matrix.org
)

If a space separated list is given for `flow.cylc[scheduling][queues][<queue name>]members`, rather than the required comma-separated list, cylc interprets the value as a single-element list containing a string with spaces. `cylc validate` reports the flow is valid and no errors appear in `cylc play` or `cylc log`, but cylc behaves as though the internal queue defined in `flow.cylc[scheduling][queues][<queue name>]` did not exist, ignoring any limits defined in it.

**Release version(s) and/or repository branch(es) affected?**
Have tested 8.0.1 and 8.0.3 and saw this behavior in both versions.

**Steps to reproduce the bug**

Create a flow that defines an internal queue, and put a space-delimited list in the `members` field, e.g.:
```
[scheduling]
  cycling mode = integer
  initial cycle point = 1
  runahead limit = P10
  [[queues]]
    [[[my_queue]]]
      limit = 2
      members = task1 task2
  [[graph]]
    P1 = "task1 => task2"
[runtime]
  [[task1]]
    script = "sleep 5; echo done>status"
  [[task2]]
    script = "sleep 5; echo done>status"
```

**Expected behavior**
`cylc validate` and `cylc play` should exit with an error on the above flow, since the `members` list should be comma delimited and 'task1 task2' is not a valid task name.

**Actual behavior**
`cylc validate` reports the above flow is valid, and `cylc play` executes the flow, ignoring the queue limit.

**Additional context**
Per @oliver-sanders in the [Element thread](https://matrix.to/#/!hwZqSYihGPuhDdIzIP:matrix.org/$1670256094139282eKgpQ:matrix.org?via=matrix.org):

> members is currently a "V_STRING_LIST" which is a comma-separated list which tolerates spaces (hence no error) We could try "V_SPACELESS_STRING_LIST" which forbids spaces (creating an error).

> Suggesting:
> 
>    1. Switch to spaceless string list since spaces are not valid in task names.
>    2. Validate queue members to ensure that all listed members exist in the [runtime] section.
> 
> Unfortunately we may need to do a bit of testing to ensure we don't break too many workflows with these changes.

### PR DESCRIPTION
**Pull request by wxtim**
Closes #5260 

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` and `conda-environment.yml`.
- [x] Tests are included (or explain why tests are not needed).
- [x] `CHANGES.md` entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/XXXX.
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

### CODE REVIEW

**Comment by wxtim** [comment]
Currently running local tests to see if this branch causes breakages.

File: cylc/flow/config.py [hunk 1]
```diff
@@ -532,6 +532,16 @@ def __init__(
...
+        for queue in self.cfg["scheduling"]["queues"]:
+            for name in self.cfg["scheduling"]["queues"][queue][
+                "members"
+            ]:
+                if (
+                    name not in self.taskdefs
+                    and name not in self.cfg['runtime']
+                ):
+                    self.implicit_tasks.add(name)
+
```
Comment (oliver-sanders) [inline cylc/flow/config.py:544]: I'm not sure that **failing** for implicit tasks in the queues is a good idea. It is common to have Jinja2 switches which turn bits of graph on and off. This change would mandate matching switches to turn off members in queues too.

```ini
[scheduling]                        
    [[queues]]                      
        [[[myq]]]                   
            members = foo, bar, baz    
    [[graph]]                       
        R1 = """                    
            foo     
            bar     
{% if BAZ %}    
            baz    
{% endif %}    
        """    
     
[runtime]    
    [[foo, bar]]    
{% if BAZ %}    
    [[baz]]    
{% endif %}  
```

There may be value in warning about implicit tasks in queues e.g. to help spot typos.

Maybe bump this bit into a "question" issue for "8.2.0".

**Review by oliver-sanders — COMMENTED** [review]
The spacelist list change works a treat.

I'm not so sure about checking for implicit tasks in the queues as this will break a lot of existing workflows and makes more programmatic workflows trickier to implement.

**Comment by oliver-sanders** [comment]
Could you rebase onto 8.1.x please.

File: CHANGES.md [hunk 2]
```diff
@@ -11,6 +11,14 @@ updated. Only the first match gets replaced, so it's fine to leave the old
 ones in. -->
 
 -------------------------------------------------------------------------------
+## __cylc-8.1.1 (<span actions:bind='release-date'>Coming Soon</span>)__
+
+### Fixes
+
+[#5312](https://github.com/cylc/cylc-flow/pull/5312) - Force queues to be
+comma separated lists only. Tasks listed in queues will be treated as implicit
+in the same way as tasks in listed in runtime.
```
Comment (hjoliver) [inline CHANGES.md:20]: I think this description is a bit misleading. Is this more accurate?:
```suggestion
[#5312](https://github.com/cylc/cylc-flow/pull/5312) - task names must be comma-separated in queue member lists. Any implicit tasks (i.e. with no task definition under runtime) assigned to a queue will generate a warning.
```
(Taking into account the suggestion below not to fail implicit tasks assigned to a queue).

File: cylc/flow/config.py (hunk 1 above)
Comment (hjoliver) [reply cylc/flow/config.py:544]: Agreed, a warning could be helpful, but failing implicit tasks here would create some pain.

File: tests/integration/test_config.py [hunk 3]
```diff
@@ -282,3 +283,39 @@ def test_parse_special_tasks_families(flow, scheduler, validate, section):
             'foo(P1D)',
             'foot(P1D)'
         }
+
+
+def test_queue_treated_as_implicit(flow, validate):
+    """Tasks listed in queue should be regarded as implicit
```
Comment (hjoliver) [inline tests/integration/test_config.py:289]: I'm not sure this docstring makes sense. It's more that any implicit task - i.e. a task name with no definition under runtime - assigned to a queue will cause validation failure (or, better, just a warning as suggested above).

**Review by hjoliver — CHANGES_REQUESTED** [review]

File: cylc/flow/config.py (hunk 1 above)
Comment (wxtim) [reply cylc/flow/config.py:544]: I thought this was overkill at the moment, but the ticket makes it look like it's a should, not a  "could sometime down the line".

**Review by oliver-sanders — APPROVED** [review]
:+1:

File: tests/integration/test_config.py [hunk 4]
```diff
@@ -282,3 +283,42 @@ def test_parse_special_tasks_families(flow, scheduler, validate, section):
...
+    assert (
+        'Queues contain tasks not defined in runtime'
+        in caplog.records[0].message
+    )
+
+
+def test_queue_treated_as_comma_separated(flow, validate):
+    """Tasks listed in queue should be separated with commas, not spaces.
+
+    https://github.com/cylc/cylc-flow/issues/5260
```
Comment (datamel) [inline tests/integration/test_config.py:312]: Not sure both these links to the issue are needed?

**Review by datamel — APPROVED** [review]
I have reproduced the issue on master and checked out this and can confirm it fixes both of the problems. Confirmed tests are also passing locally for me. Thanks @wxtim.

**Review by datamel — COMMENTED** [review]
Actually, just to confirm the behaviour of `cylc play`... before I had manually tested only `validate`. Cylc play is continuing for me when I have a space separated list, including an item which isn't in the runtime. I do get the warning in the log, but it keeps going. Are we expecting a failure here instead? 
Removing my approval for now until I understand desired behaviour.

**Comment by oliver-sanders** [comment]
Hmmm, that doesn't sound right.

**Comment by oliver-sanders** [comment]
I can't get it to actually play the workflow, however, the play error only comes out after the scheduler has started due to the way validation is performed:

```
[qval] $ cylc val .                                                                
ListValueError: (type=list) [scheduling][queues][myq]members = foo bar - (list item   "foo bar" cannot contain a space character)
[qval] $ cylc vip .    
$ cylc validate ~/cylc-tmp/qval
ListValueError: (type=list) [scheduling][queues][myq]members = foo bar - (list item   "foo bar" cannot contain a space character)
[qval] $ cylc install .
INSTALLED qval/run1 from ~/cylc-tmp/qval
[qval] $ cylc play qval/run1 --no-detach
                  
   ▪ ■  Cylc Workflow Engine 8.1.1.dev
   ██   Copyright (C) 2008-2023 NIWA
  ▝▘    & British Crown (Met Office) & Contributors
                  
2023-01-25T10:16:46Z INFO - Extracting job.sh to ~/cylc-
    run/qval/run1/.service/etc/job.sh
2023-01-25T10:16:46Z INFO - Workflow: qval/run1
2023-01-25T10:16:46Z ERROR - Workflow shutting down - ListValueError:
    (type=list) [scheduling][queues][myq]members = foo bar - (list item "foo bar"
    cannot contain a space character)
2023-01-25T10:16:46Z INFO - DONE
```

**Comment by datamel** [comment]
I'll try to reproduce, I may have done something wrong.

**Review by datamel — APPROVED** [review]
I must have been mistaken, working as expected! Sorry!!

**Comment by oliver-sanders** [comment]
Changing PR base to 8.1.x, rebase may be required

**Comment by oliver-sanders** [comment]
Got away without rebase, please merge when the test pass.

File: tests/integration/test_config.py (hunk 4 above)
Comment (wxtim) [reply tests/integration/test_config.py:312]: I think it's worth adding them to the tests they go with.

**Pull request merged**